src/KubernetesClient/Watch.cs

```diff
@@ -0,0 +1,482 @@
+        public WatchReader(KubernetesResponse response) : this((response ?? throw new ArgumentNullException(nameof(response))).Message) { }
+
+        /// <summary>Initializes a new <see cref="WatchReader{T}"/> from a <see cref="HttpResponseMessage"/> containing a watch stream.</summary>
+        public WatchReader(HttpResponseMessage response)
+        {
+            Response = response ?? throw new ArgumentNullException(nameof(response));
+            if (response.Content == null && !IsError) throw new ArgumentException("The response has no content.");
+        }
+
+        /// <summary>Initializes a new <see cref="WatchReader{T}"/> from a <see cref="Stream"/> containing a watch stream.</summary>
+        public WatchReader(Stream stream) => CreateReader(stream ?? throw new ArgumentNullException(nameof(stream)));
+
+        /// <summary>Indicates whether the request failed and so the <see cref="Response"/> does not contain a valid watch stream.</summary>
+        public bool IsError => Response != null && (int)Response.StatusCode >= 400;
+
+        /// <summary>Gets the underlying <see cref="HttpResponseMessage"/>, or null if the watch reader was initialized from a <see cref="Stream"/>.</summary>
+        public HttpResponseMessage Response { get; }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            streamReader?.Dispose();
+            Response?.Dispose();
+        }
+
+        /// <summary>Reads the next <see cref="WatchEvent{T}"/> from the stream, or returns null if the stream was closed or
+        /// <see cref="IsError"/> is true.
+        /// </summary>
+        public async Task<WatchEvent<T>> ReadAsync(CancellationToken cancelToken = default)
+        {
+            if (IsError) return null;
+            cancelToken.ThrowIfCancellationRequested();
+            if (reader == null) CreateReader(await Response.Content.ReadAsStreamAsync().ConfigureAwait(false));
+            try
+            {
+                streamReader.CancelToken = cancelToken;
+#if NETCOREAPP2_1
+                if (!await reader.ReadAsync(cancelToken).ConfigureAwait(false)) return null; // wait for the next event
+#else
+                if (!reader.Read()) return null; // wait for the next event
+#endif
+            }
+            catch (IOException ex) when (ex.InnerException is WebException wex && wex.Status == WebExceptionStatus.RequestCanceled)
+            {
+                throw new OperationCanceledException("The watch read was canceled.", ex.InnerException, cancelToken);
+            }
+
+            // deserialize the watch event
+            var e = new WatchEvent<T>();
+            bool gotType = false;
+            while (true)
+            {
+                reader.Read(); // move to the next property, if any
+                if (reader.TokenType != JsonToken.PropertyName) break;
+                string name = (string)reader.Value;
+                reader.Read(); // move to the property value
```

Comment:
Is is safe to ignore the return value of `Read`? There's no guarantee the property value can be read and it would (bizarrely) result in an `ArgumentException`:
```csharp
var r = new JsonTextReader(new StringReader("{\"type\":"));
Read();
Read();
Read();
Enum.Parse(typeof(WatchEventType), (string)r.Value, true);
void Read() => Console.WriteLine($"Read={r.Read()}, {r.TokenType}={r.Value}");
// Read=True, StartObject=
// Read=True, PropertyName=type
// Read=False, PropertyName=type
// System.ArgumentException: Requested value 'type' was not found.
```

Reply:
I'll improve it.